### ISSUE TITLE
Updating ipynb using Notebook interface doesn't trigger update of RenderMime extensions

### ISSUE THREAD

**Issue by tobinjones** [jupyterlab/jupyterlab#4547]
I'm trying to develop a JupyterLab extension that uses the `IRenderMime.IExtension` interface to provide a alternative viewer for ipynb notebook files. The extension runs the notebook through NBConvert and displays the output.

The documentation for `renderModel(model: IMimeModel)` says:

> This method may be called multiple times during the lifetime
> of the widget to update it if and when new data is available.

If I open a notebook document using the plaintext "Editor", and also open it with my "NBConvert" extension, this works as expected: If I edit the notebook in the plaintext "Editor" and then save, then `renderModel()` is called on my extension.

However, if I open the notebook using the standard "Notebook" interface, and save -- `renderModel()` is not called. Even if I cause `renderModel()` to be called by re-arranging the notebook interface, the data provided is stale. Only by closing *all* alternate renderers of the notebook and then re-opening them, does my extension see the changes from disk.

You can reproduce this with just vanilla JupyterLab by:

1. Open an `.ipynb` file in the default "Notebook" interface
2. Right click on the same file and choose "Open With">"JSON"
3. Make a change to the notebook in the "Notebook" interface, and save it.
4. Observe that the "JSON" renderer is showing stale data

I think this bug is perhaps related to #1709, but I'm not deep enough in the project to know exactly where it's coming from or how hard it might be to fix. Does anyone have guidance? Thanks!

**Comment by jasongrout** [comment]
This is a hard one to fix. Basically, the notebook viewer uses a different server request to get a notebook, since the server adds extra information to notebook files. That means that the models in the frontend are different for an editor viewer and the notebook viewer for the same file. Basically, in order to see changes from one to the other, the change has to be saved to disk, and the other viewer has to recognize that the file has been updated and revert to its state.

The proper fix here is to move the models to the server side, which is in the works but will take a while (hopefully before 1.0, though).

For now, if you want your extension to work more naturally with the notebook viewer, you can use the notebook model instead of the text editor model (that way you'll share the model the notebook uses naturally). To do this, register your extension with the notebook file type.

**Comment by tobinjones** [comment]
Thanks so much for the explanation, I understand how things came to be this way, and it's great to hear that there's a fix in the works!

I don't understand your suggestion about the "notebook model" though, sorry. I currently instantiate my extension with something like:

``` typescript
const extension: IRenderMime.IExtension = {
  id: 'jupyterlab_nbconvert_renderer:plugin',
  rendererFactory,
  rank: 99,
  dataType: 'json',
  documentWidgetFactoryOptions: [{
    name: 'NBConvertView',
    primaryFileType: 'notebook',
    fileTypes: ['notebook']
  }]
};
```

Can you please elaborate on what you mean by "register your extension with the notebook file type"?

**Comment by tobinjones** [comment]
For the moment, I'll close this -- I can work around by dropping the "IRenderMime.IExtension" convenience, and just making a "normal" lab extension.

There seems to be a comment about the notebook refactor that will ultimately fix this in #2540 -- I'll follow there for updates.
